### PR TITLE
fix: manual component upgrades/downgrades

### LINKED ISSUE #248: Providing image version in GatewayConfiguration without containerImage doesn't work
### Current Behavior

Currently if you deploy the following:

```yaml
kind: GatewayConfiguration
apiVersion: gateway-operator.konghq.com/v1alpha1
metadata:
  name: kong
  namespace: default
spec:
  dataPlaneDeploymentOptions:
    containerImage: kong/kong
    version: 2.7.0
---
kind: GatewayClass
apiVersion: gateway.networking.k8s.io/v1alpha2
metadata:
  name: kong
spec:
  controllerName: konghq.com/gateway-operator
  parametersRef:
    group: gateway-operator.konghq.com
    kind: GatewayConfiguration
    name: kong
    namespace: default
---
kind: Gateway
apiVersion: gateway.networking.k8s.io/v1alpha2
metadata:
  name: kong
  namespace: default
spec:
  gatewayClassName: kong
  listeners:
  - name: http
    protocol: HTTP
    port: 80
```

This works and results in a `Deployment` for a `DataPlane` deployed with the correct image: `kong/kong:2.7.0`.

However, if you don't provide the image:

```yaml
kind: GatewayConfiguration
apiVersion: gateway-operator.konghq.com/v1alpha1
metadata:
  name: kong
  namespace: default
spec:
  dataPlaneDeploymentOptions:
    version: 2.7.0
---
kind: GatewayClass
apiVersion: gateway.networking.k8s.io/v1alpha2
metadata:
  name: kong
spec:
  controllerName: konghq.com/gateway-operator
  parametersRef:
    group: gateway-operator.konghq.com
    kind: GatewayConfiguration
    name: kong
    namespace: default
---
kind: Gateway
apiVersion: gateway.networking.k8s.io/v1alpha2
metadata:
  name: kong
  namespace: default
spec:
  gatewayClassName: kong
  listeners:
  - name: http
    protocol: HTTP
    port: 80
```

Then the version will be ignored and the default version will be used instead.

### Expected Behavior

When `version` is provided for a component, even if `containerImage` is _not_ the version should be applied to the default image.

### Steps To Reproduce

1. `ktf env create --addon metallb`
2. `kubectl kustomize https://github.com/Kong/kubernetes-ingress-controller/config/crd | kubectl apply -f -`
3. `kubectl kustomize https://github.com/kubernetes-sigs/gateway-api/config/crd | kubectl apply -f -`
4. `kubectl kustomize https://github.com/kong/gateway-operator/config/default | kubectl apply -f -`
5. `kubectl -n kong-system wait --for=condition=Available=true --timeout=120s deployment/gateway-operator-controller-manager`
6. `kubectl apply -f <yaml provided above>.yaml`
7. `kubectl get pod <pod> -o yaml | grep image`

### Kong Gateway Operator version

24ecefc85db7e0bfffbced38c76d74ebedc6885b

### Kubernetes version

```shell
1.22.0
```

### Additional Notes

We should ensure we make this work properly for both `DataPlane`  _and_ `ControlPlane`, though the examples only demonstrated `DataPlane`.

### LINKED ISSUE #249: Manual upgrades not working with GatewayConfiguration
### Current Behavior

When defining a custom `containerImage` and `version` for the `dataplaneDeploymentOptions`, if I update the `version` to a later version the upgrade is never triggered and the older version `Pod` remains.

### Expected Behavior

When upgrading the `version` the `Deployment` should be updated accordingly.

### Steps To Reproduce

Deploy with gateway 2.7.0:

```yaml
kind: GatewayConfiguration
apiVersion: gateway-operator.konghq.com/v1alpha1
metadata:
  name: kong
  namespace: default
spec:
  dataPlaneDeploymentOptions:
    containerImage: kong/kong
    version: 2.7.0
---
kind: GatewayClass
apiVersion: gateway.networking.k8s.io/v1alpha2
metadata:
  name: kong
spec:
  controllerName: konghq.com/gateway-operator
  parametersRef:
    group: gateway-operator.konghq.com
    kind: GatewayConfiguration
    name: kong
    namespace: default
---
kind: Gateway
apiVersion: gateway.networking.k8s.io/v1alpha2
metadata:
  name: kong
  namespace: default
spec:
  gatewayClassName: kong
  listeners:
  - name: http
    protocol: HTTP
    port: 80
```

Wait for the `Pod` to be full provisioned and the `Gateway` to be responding on it's `LoadBalancer` IP, and then use:

```console
kubectl edit gatewayconfigurations kong
```

To update the version to `2.8.0`.

The `Deployment` wont be upgraded.

### Kong Gateway Operator version

24ecefc85db7e0bfffbced38c76d74ebedc6885b

### Kubernetes version

```shell
v1.22.0
```

### Anything else?

We should ensure that this is fixed for both `DataPlane` _and_ `ControlPlane`.

It's worth noting, that the gateway controller _is updating the `DataPlane`_ when this happens, but the dataplane controller is not updating its `Deployment`.

### PR DESCRIPTION
**Pull request by shaneutt**
**What this PR does / why we need it**:

Manual version changes were broken, the `ControlPlanes` and `DataPlanes` would get updated from a `GatewayConfiguration`, however the underlying `Deployments` never got those updates. This patch fixes that and makes it possible to easily upgrade and downgrade manually using the `GatewayConfiguration` (or `ControlPlane`/`DataPlane` directly).

**Which issue this PR fixes**

Resolves #249

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -106,6 +106,45 @@ Guides][kic-guides] for more information).
...
+  dataPlaneDeploymentOptions:
+    containerImage: kong/kong
+    version: 2.7.0
+  controlPlaneDeploymentOptions:
+    containerImage: kong/kubernetes-ingress-controller
+    version: 2.4.2
+```
+
+The above configuration will deploy all `DataPlane` resources connected to the
+`GatewayConfiguration` (by way of `GatewayClass`) using `kong/kong:2.7.0` and
```
Comment (pmalek) [inline README.md:131]: I've seen way of connecting `GatewayClass` with `GatewayConfiguration` in the description of #249 but I'm not sure if we have that documented somewhere. Do we? If not can we add an issue for adding this a doc about how to go about linking those 2?

File: controllers/utils.go [hunk 2]
```diff
@@ -339,3 +340,62 @@ func addLabelForOwner(obj client.Object, owner client.Object) {
...
+func ensureContainerImageUpdated(container *corev1.Container, image *string, version *string) (updated bool, err error) {
+	imageParts := strings.Split(container.Image, ":")
+	if len(imageParts) > 2 {
+		err = fmt.Errorf("invalid container image found: %s", container.Image)
+		return
+	}
+
+	switch {
+	// if both the image and the version were provided, we expect the container'
+	// image to match exactly.
```
Comment (pmalek) [inline controllers/utils.go:361]:
```suggestion
	// if both the image and the version were provided, we expect the container's
	// image to match exactly.
```

File: controllers/utils.go [hunk 3]
```diff
@@ -339,3 +340,62 @@ func addLabelForOwner(obj client.Object, owner client.Object) {
...
+		expectedImageAndVersion := fmt.Sprintf("%s:%s", *image, *version)
+		if container.Image != expectedImageAndVersion {
+			container.Image = expectedImageAndVersion
+			updated = true
+		}
+	// if only the image was provided we expect the image to match but we don't
+	// worry about what the version is.
+	case image != nil && *image != "":
+		expectedImage := *image
+		parts := strings.Split(container.Image, ":")
```
Comment (pmalek) [inline controllers/utils.go:372]: This is basically `imageParts` which is already available (assigned in line 353), right? So we can drop this and just reuse `imageParts` if I read this right.

File: controllers/utils.go [hunk 4]
```diff
@@ -339,3 +340,62 @@ func addLabelForOwner(obj client.Object, owner client.Object) {
...
+			if container.Image != expectedImage {
+				container.Image = expectedImage
+				updated = true
+			}
+		}
+	// if only the image version was provided we expect the tag to match but we
+	// don't worry about what the base image is.
+	case version != nil && *version != "":
+		expectedVersion := *version
+		parts := strings.Split(container.Image, ":")
```
Comment (pmalek) [inline controllers/utils.go:388]: Ditto here?

**Review by pmalek — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (shaneutt) [reply README.md:190]: We currently only document it as a sample manifest: https://github.com/Kong/gateway-operator/blob/main/config/samples/gateway.yaml

But I've now also added documentation in the README (which I think is getting pretty big, but I figure for now we can keep filling up the README and then start moving stuff out of there once we have a spot on docs.konghq.com) in 596d8d71cca6018cba2f25f704c3c2c5901d6ec1.

File: controllers/utils.go (hunk 3 above)
Comment (shaneutt) [reply controllers/utils.go]: Oh good call, this was an oversight during refactoring. Fixed in 596d8d71cca6018cba2f25f704c3c2c5901d6ec1 :+1:

File: controllers/utils.go (hunk 4 above)
Comment (shaneutt) [reply controllers/utils.go]: Fixed in 596d8d71cca6018cba2f25f704c3c2c5901d6ec1 :+1:

**Review by pmalek — DISMISSED** [review]
🚀

File: README.md [hunk 5]
```diff
@@ -106,6 +106,104 @@ Guides][kic-guides] for more information).
...
+
+With the `parametersRef` in the above `GatewayClass` being used to attach the
+`GatewayConfiguration`, that configuration will start applying to all `Gateway`
+resources created for that class, and will retroactively apply to any `Gateway`
+resources previously created.
+
+[kic]:https://github.com/kong/kubernetes-ingress-controller
+[gw]:https://github.com/kong/kong
+
+#### Gateway Upgrades/Downgrades
```
Comment (pmalek) [inline README.md:168]: BTW: any particular reason why `Upgrades` and `Downgrades` use this capitalization? 🤔

**Review by pmalek — APPROVED** [review]

File: README.md (hunk 5 above)
Comment (shaneutt) [reply README.md:168]: No reason: 
```suggestion
#### Gateway upgrades/downgrades
```

**Review by pmalek — APPROVED** [review]

**Pull request merged**